ODBC backend plugin

flo91 opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

The read-access to ODBC data sources is working, unit tests are present and a new command `kdb mountOdbc` was added to provide basic mount functionality for ODBC data sources.

**Work in progress / open:**
- implement write functionality --> will be done in new PR
- implement support for binary values --> will be done in new PR

**Done:**
- implement metadata support
- offer more configuration options (e.g. table- and column-names)
- remove hardcoded mountpoint for the ODBC backend
- update tooling to allow the creation of user-defined mountpoints for the ODBC backend
  - new `kdb mountOdbc` command added
- extend documentation
  - comprehensive Tutorial written
- add tests

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [x] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
-->

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

tmakar reviewed (commented):
I thought I will do a first check on your draft PR , hope you don’t mind ;)

LGTM, would have maybe split some things into separate functions instead of having comments with steps, but this is a choice of preference I think.

flo91 commented:
jenkins build libelektra please

flo91 commented:
jenkins build libelektra please

atmaxinger commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,827 @@
+			SQLFreeHandle (SQL_HANDLE_STMT, sqlStmt);
+			return NULL;
+		}
+		else
```
You can remove the else here to reduce nesting.

flo91 replied:
Good catch! 👍🏻

atmaxinger commented on src/plugins/backend_odbc/sampleDb/prepareDB.sql:
```diff
@@ -0,0 +1,30 @@
+);
+
+CREATE TABLE metaKeys (
+    keyName TEXT,
```
```suggestion
    keyName TEXT NOT NULL,
```

flo91 commented on doc/tutorials/odbc-backend.md:
```diff
@@ -0,0 +1,369 @@
+### Preparing the SQLite database
+
+As SQLite as a file-based DBMS, we first create a file for the new database.
+Afterward, we execute the SQL-statements to create the tables and insert some tuples.
```
I'm also used to "afterwards" and initially used the word.
But CLion detected a spelling error and I checked https://dict.leo.org/german-english/afterward and https://grammarist.com/usage/afterward-afterwards.
It tells that "afterward" is more common in American English.
As [American English](https://www.libelektra.org/devgettingstarted/documentation#style) should be used for Elektra, I used "afterward".

flo91 commented on src/plugins/backend_odbc/backend_odbc_helpers.c:
```diff
@@ -0,0 +1,533 @@
+		{
+			if (i)
+			{
+				result[dsnCount] = (char *) dsn;
```
Thanks! 👍🏻 
Some things there definitely went wrong while refactoring.
It was one of the first functions I wrote while studying the ODBC API and is currently not in use. It should be added in some way to the `kdb` tool.

flo91 commented on src/plugins/backend_odbc/backend_odbc.c:
```diff
@@ -0,0 +1,173 @@
+	}
+	case ELEKTRA_KDB_GET_PHASE_CACHECHECK:
+		/* TODO: implement cache */
+		// return ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
```
Basically to get rid off the Clang-Tidy warning "Switch has 2 consecutive identical branches".
But yes, it seems safer to uncomment the line and return here explicitly.

flo91 commented on src/plugins/backend_odbc/backend_odbc.c:
```diff
@@ -0,0 +1,173 @@
+		return ELEKTRA_PLUGIN_STATUS_ERROR;
+	}
+
+	elektraPluginSetData (plugin, dsConfig);
```
Are you sure that this would be safe in all cases?
Is it guaranteed that this memory can be freed?
Is it sure that it was allocated with malloc or calloc and is afterwards not accessed via another pointer somewhere else.

flo91 commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,827 @@
+	    !(dsConfig->metaTableKeyColName) || !(*(dsConfig->metaTableKeyColName)) || !(dsConfig->metaTableMetaKeyColName) ||
+	    !(*(dsConfig->metaTableMetaKeyColName)) || !(dsConfig->metaTableMetaValColName) || !(*(dsConfig->metaTableMetaValColName)))
+	{
+		return "";
```
The idea was to indicate via the return value if a memory allocation failed (`NULL` is returned) or if data to create the query string is missing (empty string is returned).
So, the caller has to check for both cases (NULL or empty) before using (and freeing) the returned value. The situation is also described in the Doxygen comment for the function.

This distinction is used here: https://github.com/ElektraInitiative/libelektra/blob/9adaca1076eea1ae2ad8280c16b81f00b7056f0e/src/plugins/backend_odbc/backend_odbc_get.c#L247-L264


But you have a point here. Maybe it's safer and more flexible to add an argument for passing the `errorKey` to the function `getSelectQueryString` and let it set the error itself, esp. if you consider that the empty string is the only case where a call to free would cause trouble (`free (NULL)` just does nothing).

flo91 replied:
That's likely the best approach.
I've removed the check and added a precondition as you suggested.

flo91 commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,827 @@
+	}
+
+	/* Build the string */
+	char * strEnd = stpcpy (queryString, "SELECT ");
```
I thought so too, but at least the ODBC-API doesn't allow parameters for the column list of a SELECT-query. :(
In the [ODBC documentation](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/statement-parameters) is written:
> Parameters are valid only in certain places in SQL statements. For example, they are **not allowed in the select list (the list of columns to be returned by a SELECT statement)**, nor are they allowed as both operands of a binary operator such as the equal sign (=), because it would be impossible to determine the parameter type.

So the escaping of the identifiers has to happen somewhere else and independent of the used ODBC driver (and therefore data source like different DBMS or also file-based formats).

One option would be to make quoted identifiers obligatory and reject parameters that already contain quote-characters.

flo91 replied:
I've added the check for detecting `SQL_IDENTIFIER_QUOTE_CHAR` in the identifiers in `dsConfig` and made using `SQL_IDENTIFIER_QUOTE_CHAR` obligatory.

The concept with the SQL-queries instead of table- and column names is an idea for a follow-up PR, but not in the scope of this PR now.

flo91 commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,827 @@
+		return NULL;
+	}
+
+	if (!setLoginTimeout (sqlConnection, 5, errorKey))
```
Certainly a good idea. 😀 
Maybe adding it as an optional value to the data source config struct is the best option.
So if you know that an ODBC data sources connects to a DBMS with a high latency, you can set a higher timeout for that specific data source.

flo91 commented on src/plugins/backend_odbc/backend_odbc_general.c:
```diff
@@ -0,0 +1,357 @@
+	}
+	else if (ret == SQL_SUCCESS_WITH_INFO)
+	{
+		setOdbcError (SQL_HANDLE_DBC, sqlConnection, "endTransaction", true, errorKey);
```
Thanks, that's a great tip! :grinning:

flo91 commented on doc/tutorials/odbc-backend.md:
```diff
@@ -0,0 +1,369 @@
+Description=General ODBC for PostgreSQL
+Driver=/usr/local/lib/psqlodbca.so
+Setup=/usr/lib/libodbcpsqlS.so
+Setup64=/usr/libd64/libodbcpsqlS.so
```
No, this was a typo.
Good catch!

flo91 commented on doc/tutorials/odbc-backend.md:
```diff
@@ -0,0 +1,369 @@
+
+On most systems, these files should be stored at `/etc/unixODBC/`.
+
+Unfortunately, the current Elektra plugins for .ini-files don't support the exact syntax that is used by
```
If I try to insert the entries as normal keys, I get the `=` char in the file.

kdb:
`kdb set user:/odbc/[SQLite] ""`


ni:
`\[SQLite] = `

mini:
`[SQLite]=`

simpleini:
`[SQLite] = 
`

On the other hand, if I try to insert the entries belonging to a section as metadata, I get the 'meta:/' prefixes in the file.

kdb:
`kdb meta-set user:/odbc/SQLite Description "SQLite Database"`

ni:
```
SQLite = 

[SQLite]
 meta:/Description = SQLite Database
```

mini:
does not support metadata!

simpleini:
`[SQLite] = `

So simpleini doesn't throw an error, but also doesn't write the metadata to the file.

If you have an idea how to get the required syntax in the ODBC .ini-files, I'd be glad if you can tell it to me.

flo91 replied:
But I've changed the tutorial now to use the more generic `line` plugin for creating the ODBC .ini-files.

flo91 commented:
@atmaxinger 
I've now addressed the issues that you've mentioned in your review and did some other bugfixes.
Additionally, I also wrote a manpage for the `kdb mountOdbc` command.

kodebach commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,842 @@
+
+				/* Check key-name column */
+				bool retGetLongData = true;
+				if (buffers->nameLenInd == SQL_NO_TOTAL)
+				{
+					/* No information about the total length is available --> The new buffer maybe must be
+					 * resized multiple times */
+					longKeyName = (char *) elektraMalloc (sizeof (char) * (KEYNAME_BUFFER_SIZE * 2));
+					retGetLongData =
+						getLongData (sqlStmt, 1, SQL_C_CHAR, &longKeyName, KEYNAME_BUFFER_SIZE * 2, errorKey);
+				}
+				else if (KEYNAME_BUFFER_SIZE <= buffers->nameLenInd)
+				{
+					longKeyName = (char *) elektraMalloc (sizeof (char) * (buffers->nameLenInd + 1));
+					retGetLongData =
+						getLongData (sqlStmt, 1, SQL_C_CHAR, &longKeyName, (buffers->nameLenInd + 1), errorKey);
+				}
```
Why not do this distinction inside `getLongData`?

flo91 replied:
see above

kodebach commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,842 @@
+		{
+			/* Create new key */
+			prevKey = curKey;
+			curKey = keyDup (errorKey, KEY_CP_NAME);
```
If the keys is used for things other than errors, it shouldn't be called `errorKey`

kodebach commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,842 @@
+				}
+
+				/* Check if the keyname changed since the last iteration */
+				if (longKeyName && prevKeyName && elektraStrCmp (longKeyName, prevKeyName) == 0)
```
That seems really inefficient. But I don't have a suggestions for a better solution (other than N+1 `SELECT`).

kodebach reviewed (changes requested):
In general this looks decent solution. However, there are some issues with this PR:

1. I really would have preferred, if this was done in multiple PRs over a longer period of time. There are some IMHO not ideal design things that can't really be changed now (what does the mountpoint definition look like). The PR also doesn't have that many files, but the code is very dense and it takes a while to review. Would have been much easier to review, if we first had a review round for the technical design.
2. More shell tests actually showing that everything works would be nice. Also you say the backend is tested with Postgres, but I don't think there is a CI test for that. So please either remove that from the docs, or add some CI tests. Without CI tests, we cannot guarantee that it keeps working in the future.
3. You use "should" in many places in the docs (e.g. "should be a primary key"). IMO it is pretty unclear what this means. What happens if I ignore this requirement? Is there a check and an error is thrown? If so the docs should say "must". If there is no check, the question is still: What happens? Also why is it better with the requirement? In a few cases (e.g. foreign key), I quite certain there isn't actually a requirement and things would work just fine. Please be clear, what things are _requirements_ and which are just _recommendations_ **and** why these are recommended/required.

I also haven't look at the C++ code for `kdb` at all, because it will hopefully be replaced by a C version soon. I assume it works, but AFAICT there are no shell tests for it, so who knows.

flo91 commented on src/plugins/backend_odbc/backend_odbc_get.c:
```diff
@@ -0,0 +1,842 @@
+		}
+
+		getDataRet = SQLGetData (sqlStmt, colNumber, targetType, (*targetValue) + (iteration * (bufferSize - 1)), bufferSize,
+					 &getDataLenOrInd);
```
Hm, the distinction is made via the return-value.
If `SQL success` is returned, the buffer was big enough.
The function is primary a helper for unknown value sizes, otherwise, just pass the known size.

The `nameLenOrInd' values inside the `struct columnData` are used in various places, so I used them for this case too to be consistent.

As the function is `static` and therefore only used internally, I think we can retain the current logic.

flo91 reviewed (commented):
Thank you so much for the very detailed and helpful review! :grinning:

I planned to add more shell tests in the PR for the write functionality,
so that complete scenarios (CRUD) can be tested.

As you've mentioned, the `kdb mountOdbc` command is intended as a temporary
solution to make to ODBC backend useful for end users.
I didn't want to invest too much time in this part, as it is going to get replaced with a C-based implementation anyway.
However, even if the number and definition of the arguments is not optimal, it does its job correctly and enables user to create custom mountpoints for ODBC data sources.

I've applied your suggestions and pushed them to this PR.

Now that I've got two good and helpful reviews from @atmaxinger and @kodebach and applied the corrections, I can finally assign this PR to be reviewed by @markus2330. 🎉

kodebach commented on doc/tutorials/odbc-backend.md:
```diff
@@ -0,0 +1,395 @@
+If you want to use `CHAR` or `VARCHAR` columns, it's your responsibility to make sure that the stored values don't exceed the maximum defined length.
+
+It is allowed that the tables contain more columns, these additional columns are not processed by Elektra,
+_must not_ be part of a PK and _must_ support NULL-values (if you want to add new keys in a future version of Elektra).
```
> If multiple rows contain the same key-name column, they are treated as one key.
The value would be taken from the first key, the metadata of the two keys would be appended.

Please add this to the docs.

flo91 replied:
Ok, I've added it, but also mentioned that this behavior is considered a part of the internal algorithm of the plugin and could change in the future.
So to be sure to get predictable and reasonable results, it's strongly advised to respect the PK and FK constraints.

kodebach replied:
> that this behavior is considered a part of the internal algorithm of the plugin and could change in the future

That's fine, I'd also have accepted if the docs just explicitly state e.g. "duplicate key names are undefined behaviour". Just the implicitly undefined behaviour made it unclear whether there is or isn't consistent handling for this or if it could even lead to crashes/errors or corrupted data.

kodebach commented on src/plugins/backend_odbc/README.md:
```diff
@@ -0,0 +1,69 @@
+The tables may also contain other columns, but they are not processed by this plugin and must support NULL- or DEFAULT-values,
+if you want to add tuples to the table via Elektra (e.g. by calling `kdb set <key> <value>`).
+
+The following properties are available:
```
IMO it's still not clear where these properties must be set. Is this an ODBC thing? Do they have to be defined in an ODBC config file?

AFAICT these can all be specified as arguments to `kdb mountOdbc`, if you wanted to say that, then the term "properties" is confusing IMO.

Also the README doesn't describe the mountpoint definition stored at  `system:/elektra/mountpoints/<MP>/definition` (and I assume created by `kdb mountOdbc`) at all. AFAICT these are the same "properties" but with slightly different names. Maybe it makes the most sense to just specify what the mountpoint definition has to look like.

flo91 replied:
Ok, now I get what you mean.
I've now described the keys that are part of the definition of an ODBC mountpoint in the `README.md` of the plugin and referenced the man page of the `kdb mountOdbc` command where the arguments of that command are described.

markus2330 commented:
@flo91 great work, looking forward to follow-up. Didn't test it yet.

@kodebach thx for the review!

@hannes99 please rework arguments of `mount odbc` as discussed (also make `odbc` a subcommand as suggested by @kodebach).
